refactor(syntax)!: remove `SymbolFlags::ArrowFunction`

overlookmotel opened the pull request:
`SymbolFlags::ArrowFunction` is an oddity, as whether a symbol is an arrow function is not statically knowable. In the following cases, `f` symbol did not have `ArrowFunction` flag set:

```js
const {f} = {f: () => {}};
```

```js
let f = 123;
f = () => {};
```

`SymbolFlags::ArrowFunction` is therefore not particularly useful, and possibly misleading. Having it complicates the transformer, and it's not used anywhere in Oxc.

This PR removes it.

overlookmotel commented:
@DonIsaac Have requested your review as I think you added this flag. But I'm not clear why - it's not used in the linter (or anywhere else in Oxc).

DonIsaac commented:
> @DonIsaac Have requested your review as I think you added this flag. But I'm not clear why - it's not used in the linter (or anywhere else in Oxc).

I did add it for the linter, but I did not refactor existing rules to use it. Removing this should be fine, as we can achieve what we need without it.

overlookmotel committed:
refactor(syntax)!: remove `SymbolFlags::ArrowFunction` (#5857)

`SymbolFlags::ArrowFunction` is an oddity, as whether a symbol is an arrow function is not statically knowable. In the following cases, `f` symbol did not have `ArrowFunction` flag set:

```js
const {f} = {f: () => {}};
```

```js
let f = 123;
f = () => {};
```

`SymbolFlags::ArrowFunction` is therefore not particularly useful, and possibly misleading. Having it complicates the transformer, and it's not used anywhere in Oxc.

This PR removes it.
